File: 02_rotate90Matrix/2.c
#include <stdio.h>

#define N 3  

void rotateCounterClockwise(int mat[N][N]) {
 
    for (int i = 0; i < N; i++) {
        for (int j = i + 1; j < N; j++) {
            int temp = mat[i][j];
            mat[i][j] = mat[j][i];
            mat[j][i] = temp;
        }
    }

    for (int j = 0; j < N; j++) {
        for (int i = 0; i < N / 2; i++) {
            int temp = mat[i][j];
            mat[i][j] = mat[N - 1 - i][j];
            mat[N - 1 - i][j] = temp;
        }
    }
}

int main() {
    int mat[N][N] = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    printf("Original Matrix:\n");
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%d ", mat[i][j]);
        }
        printf("\n");
    }

    rotateCounterClockwise(mat);

    printf("\nRotated Counter-Clockwise Matrix:\n");
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            printf("%d ", mat[i][j]);
        }
        printf("\n");
    }

    return 0;
}

File: 02_rotate90Matrix/2.py
class Solution:
    def rotateMatrix(self, mat):
        n = len(mat)  
        
        for i in range(n):
            for j in range(i + 1, n):
                
                mat[i][j], mat[j][i] = mat[j][i], mat[i][j]
        for j in range(n):
            for i in range(n // 2):
                mat[i][j], mat[n - 1 - i][j] = mat[n - 1 - i][j], mat[i][j]

sol = Solution()
matrix = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
]

print("Original Matrix:")
for row in matrix:
    print(row)

sol.rotateMatrix(matrix)

print("\nRotated Counter-Clockwise Matrix:")
for row in matrix:
    print(row)
